internal/database/dbmock/dbmock.go

```diff
@@ -0,0 +1,113 @@
+// The dbmock package facilitates embedding mock stores directly in the
+// datbase.DB object.
+package dbmock
+
+import (
+	"context"
+
+	"github.com/sourcegraph/sourcegraph/internal/database"
+	"github.com/sourcegraph/sourcegraph/internal/database/basestore"
+)
+
+// Configurable is any type that, when given database.DB, turns into T.
+type Configurable[T any] interface {
+	WithDB(database.DB) T
+}
+
+// BaseStore is a store without a database connection.
+// It can be turned into a store with a database connection
+// by calling .WithDB and providing a DB.
+type BaseStore[K any, T Configurable[K]] interface {
+	WithDB(database.DB) K
+}
+
+// baseStore implements BaseStore.
+// It checks the provided database.DB for any mocks.
+type baseStore[K any, T Configurable[K]] struct {
+	store T
+}
+
+func NewBaseStore[K any, T Configurable[K]](store T) BaseStore[K, T] {
+	return &baseStore[K, T]{
+		store: store,
+	}
+}
```

Comment:
I think we could simplify all these, right?

```suggestion
// BaseStore is a store without a database connection.
// It can be turned into a store with a database connection
// by calling .WithDB and providing a DB.
type BaseStore[K any] interface {
	WithDB(database.DB) K
}

// baseStore implements BaseStore.
// It checks the provided database.DB for any mocks.
type baseStore[K any] struct {
	store Configurable[K]
}

func NewBaseStore[K any](store Configurable[K]) BaseStore[K] {
	return &baseStore[K]{
		store: store,
	}
}
```

Reply:
Brilliant! , knew I was missing something here